Rust for VGEM

mairacanal opened the pull request:
# TODO

- [x] Make the DRM device be unregistered upon removal
- [x] Add timeout feature
- [x] Are all fences being properly put?
- [x] Implement the close callback: It is unnecessary as Xarray deals with the drop for us
- [x] Is the timer being properly deallocated?
- [x] Inspect all possible error cases
- [x] Add all the SAFETY comments
- [x] Fix all the FIXME commits
- [x] Use Boqun's Timer Abstraction

# Questions

- [x] Is a mutex needed to protect Xarray? No!
     - rust/kernel/xarray.rs:293: // SAFETY: XArray is thread-safe and all mutation operations are internally locked.

- [ ] It works with seqno = 0 for some reason

# Possible Improvements

- [x] Xarray can hold a VgemFence instead of a UniqueFence<VgemFence>

# RFC

- v1: https://lore.kernel.org/dri-devel/20230317121213.93991-1-mcanal@igalia.com/T/

fbq committed:
rust: Introduce jiffies-based timer support

Signed-off-by: Boqun Feng <boqun.feng@gmail.com>

mairacanal committed:
rust: platform: add direct registration for a platform device

Allows drivers to directly allocate and register a device. Although this
style of driver is discouraged, it is useful for virtual devices that
cannot be probed. Also, handles the deallocation of a registered device
through the Drop trait.

Signed-off-by: Maíra Canal <mcanal@igalia.com>

mairacanal committed:
drm/rustgem: implement a Rust version of VGEM

The VGEM driver is a non-hardware backed GEM service and it is currently
implemented in C on the DRM subsystem. This patch introduces a new
version of the VGEM driver written in Rust. This driver provides the
same functionalities and uses the same UAPI of the original VGEM driver.

Signed-off-by: Maíra Canal <mcanal@igalia.com>

mairacanal committed:
rust: platform: add Device.coerse_dma_masks()

Allows drivers to configure the DMA masks for a device dealing with the
case where the device does not have dev->dma_mask appropriately setup.
Implemented here, not in device, because it requires a mutable platform
device reference.

Signed-off-by: Maíra Canal <mcanal@igalia.com>

mairacanal committed:
rust: device: add Device Resource Management abstraction

Create an abstraction to allow Rust drivers to open a new devres group for
the device with a defined id. Moreover, implement the related Drop trait
to release the devres group.

Signed-off-by: Maíra Canal <mcanal@igalia.com>

mairacanal committed:
drm/vgem: move IOCTLs numbers to enum

Bindgen doesn't expand C macros, which makes DRM_IOCTL_VGEM_FENCE_ATTACH
and DRM_IOCTL_VGEM_FENCE_SIGNAL not accessible through the Rust
bindings. So, move the IOCTLs numbers to a enum, preserving the same
values and the same name. This won't produce any functional changes to
the UAPI and it will make it possible to use this UAPI in Rust.

Signed-off-by: Maíra Canal <mcanal@igalia.com>

mairacanal committed:
rust: make platform device implement the Send and Sync traits

Currently, platform device holds only a pointer, which is safe to be
used from any thread. This is useful for devices that are not probed,
such as virtual devices.

Signed-off-by: Maíra Canal <mcanal@igalia.com>

mairacanal committed:
drm/rustgem: create dummy IOCTL with number 0x00

In order to declare IOCTLs with the current Rust abstractions, we use
the kernel::declare_drm_ioctls! macro. This macro considers that the
IOCTLs are in the right order and there are no gaps. This isn't the case
for vgem, which start the IOCTLs with 0x01, instead of 0x00.

With the intention to use the kernel::declare_drm_ioctls! macro, create
a dummy IOCTL with number 0x00, that returns EINVAL.

Signed-off-by: Maíra Canal <mcanal@igalia.com>

mairacanal committed:
rust: dma_resv: add DMA Reservation abstraction

The DMA reservation object provides a mechanism to manage a container
of dma_fence object associated with a resource. So, add an abstraction
to allow Rust drivers to interact with this subsystem.

Signed-off-by: Maíra Canal <mcanal@igalia.com>

mairacanal committed:
rust: drm: gem: add method to return DmaResv from GEMObject

A GEMObject, related to struct drm_gem_object, holds a pointer
to reservation object associated with the this GEM object. Therefore,
expose this reservation object through the DmaResv safe abstraction.

Signed-off-by: Maíra Canal <mcanal@igalia.com>

mairacanal committed:
rust: dma_fence: add method to return an indication if the fence is signaled

To indicate the current status of the fence, there is a kernel function that
returns an indication if the fence is signaled yet. Therefore, add a method
in the Rust abstraction to return an indication if the fence is signaled.

Signed-off-by: Maíra Canal <mcanal@igalia.com>

mairacanal committed:
rust: dma_fence: expose the fence's seqno publically

Each fence has a linear increasing sequence number inside the execution
context, that can be used to decide which fence would be signaled later.
So, expose this attribute to the Rust drivers through a method.

Signed-off-by: Maíra Canal <mcanal@igalia.com>

XX commented on drivers/gpu/drm/rust-gem/file.rs:
```diff
@@ -0,0 +1,149 @@
+        };
+
+        // Record the fence in our idr for later signaling
+        match resv.add_fences(fence.deref(), 1, usage) {
+            Ok(_) => match file.fences.alloc(Some(Box::try_new(Some(fence))?)) {
+                Ok(id) => data.out_fence = id as u32,
+                Err(_) => {}
+            },
+            Err(_) => {
+                fence.signal()?;
+            }
+        }
```
Can be simplified:
```rust
if resv.add_fences(fence.deref(), 1, usage).is_ok() {
    if let Ok(id) = match file.fences.alloc(Some(Box::try_new(Some(fence))?)) {
        data.out_fence = id as u32;
    }
} else {
    fence.signal()?;
} 
```

mairacanal committed:
drm/rustgem: implement timeout to prevent hangs

In order to prevent possible hangs, if the fence is not signaled for
more than 10 seconds, force the fence to expire by being signaled.

Signed-off-by: Maíra Canal <mcanal@igalia.com>

mairacanal committed:
rust: drm: gem: shmem: set map_wc on gem_create_object callback

Some drivers use the gem_create_object callback to define the mapping of
the object write-combined (map_wc). Currently, the DRM Rust abstractions
doesn't allow such operation. So, add a constant to the DriverObject trait
to allow drivers to set map_wc on the gem_create_object callback. By
default, the constant returns false, which is the shmem default value.

Signed-off-by: Maíra Canal <mcanal@igalia.com>

riking commented on drivers/gpu/drm/rustgem/file.rs:
```diff
@@ -0,0 +1,143 @@
+        _data: &mut bindings::drm_vgem_dummy,
+        _file: &DrmFile,
+    ) -> Result<u32> {
+        Err(EINVAL)
```
ENOTSUP seems like a more appropriate error code here
